src/impression.js

```diff
@@ -235,3 +241,65 @@ function applyResponse(win, response) {
     win.history.replaceState(null, '', newHref);
   }
 }
+
+/**
+ * Return a promise that whether appending extra url params to outgoing link is required.
+ * @param {!Window} win
+ * @return {!Promise<boolean>}
+ */
+export function shouldAppendExtraParams(win) {
+  return whenDocumentReady(win.document).then(doc => {
+    return !!doc.querySelector('amp-analytics[type=googleanalytics]');
+  });
+}
+
+/**
+ * Return the extra url params string that should be appended to outgoing link
+ * @param {!Window} win
+ * @param {!Element} target
+ * @return {string}
+ */
+export function getExtraParamsUrl(win, target) {
+  // Get an array with extra params that needs to append.
+  const url = parseUrl(win.location.href);
+  const params = parseQueryString(url.search);
+  const appendParams = [];
+  for (let i = 0; i < DEFAULT_APPEND_URL_PARAM.length; i++) {
+    const param = DEFAULT_APPEND_URL_PARAM[i];
+    if (typeof params[param] !== 'undefined') {
+      appendParams.push(param);
+    }
+  }
+
+  // Check if the param already exists
+  const additonalUrlParams = target.getAttribute('data-amp-addparams');
+  let href = target.href;
+  if (additonalUrlParams) {
+    href = addParamsToUrl(href, parseQueryString(additonalUrlParams));
+  }
+  const loc = parseUrl(href);
```

Comment:
Could this be simplified if appendParams were an object, that is merged with parseQueryString(additionalUrlParams)? And then you could call addParamsToUrl on the merged object.

Reply:
The thing is that I don't want to merge them until finally appended to the outgoing link. There're whitelist restrictions that need to be applied to `addtionalUrlParams` but not `appendParams`. Also we are having different behaviors on non-whitelisted outgoing link origin